### PR TITLE
Fix versioned release schema and merge rules

### LINKED ISSUE #770: tag should be omitWhenMerged
The behavior of OCDS Merge is to set the tag to `compiled` for compiledRelease, and to move the tag onto the versioned object for versionedRelease. In other words, the original tag field is omitted.

Assigning to @kindly to confirm. The docs don't describe the first behavior, which we can fix as part of this issue.

### LINKED ISSUE #771: Add missing documentation on merging
I don't see where in the documentation the following OCDS Merge behaviors are prescribed:

* [x] The `id` of the compiledRelease is the `id` of the latest release (#810 #834 set `{ocid}-{date}`)
* [x] The `date` of the compiledRelease is the `date` of the latest release #810 #834 
* [x] The `tag` of the compiledRelease is `["compiled"]` #770 #810
* [x] The `ocid` of the versionedRelease is not versioned (only specified in schema) #810
* [x] The `tag` of the versionedRelease is omitted #770 #810
* [x] If a value in a release is the same as in its preceding release, don't add a versioned object #810

@kindly Did I miss these somewhere?

* [x] As part of this we can clarify precedence, i.e. that `omitWhenMerged` precedes all other rules.

### LINKED ISSUE #798: Versioned release schema is inconsistent with merge rules
## Issue

The [merge rules](http://standard.open-contracting.org/latest/en/schema/merging/#versioned-data) state that all fields are versioned "with the exception of the `id` property of objects within an array". The OrganizationReference definition appears in both the context of an object in an array (e.g. `suppliers`) and in the context of a plain object (e.g. `buyer`). According to the rules, in one case its `id` property should be versioned (`buyer`, etc.), in the other it shouldn't (`suppliers`, etc.).

However, the versioned release schema uses the same definition for both contexts, which states that `id` isn't versioned. But, we *do* want the `id` in `buyer` and other single organization references to be versioned.

Classification and Identifier also appear in the context of both an object in an array (`additionalClassifications`, `additionalIdentifiers`) and in the context of a plain object (`classification`, `identifier`). However, those arrays are whole list merge, and objects in that context already have no versioned fields. That said, the versioned release schema has a bug in that it states that their `id` isn't versioned. Compared to the above, this is an easy fix.

## Discussion

**Note: Ignore this discussion, as a simpler solution has been found.**

I believe that the merge rules are correct and that the versioned release schema is incorrect. The three broad solutions are:

1. Change the versioned release schema to merge these objects as a whole, if not in an array
1. Change the versioned release schema to make `id` versioned if not in an array of objects
1. Remove the versioned release schema from the standard

### Merge the objects as a whole

We either:

1. Add a new `wholeObjectMerge` property, to indicate that the `buyer`, etc. object should be merged as a whole. (For non-`$ref` objects, we can simply add a `type` of "null" to achieve the same behavior. However, we can't do that for Classification, Identifier, and OrganizationReference, because these definitions are references from arrays, and we don't want to allow `null` in arrays.)
1. Use different definitions when referring to an object in array-context than when referring to it in non-array-context. This allows us to use the strategy of adding a `type` of "null" mentioned before, which has the advantage of not adding a new merging rule. There's some discussion in #606 about what sort of tool support we'll need to overcome the disadvantages of a bigger, repetitive schema.

### Make `id` versioned if not in an array of objects

As above, we either (1) add a new merge property to indicate that this object's `id` field should be versioned, or (2) make a copy of the definition and set `"versionId": true` for its `id` field.

### Remove the versioned release schema from the standard

This has been mentioned before, but over time I've appreciated the functionality that the versioned release provides. It makes it very easy to see changes over time, which is a common use case. That said, hardly anyone has implemented it, and it is always possible to calculate from individual releases. We also, evidently, have trouble defining its schema, documenting it, and implementing it.

## Proposal

Within the first two approaches, I prefer the second sub-option as it helps solve #606 as well, which will become an increasingly severe issue as the number of extensions increases. It's likely more work than the first option, however. The first option can be pursued in a patch release. The second option *might* be pursued in a patch release – it depends if it affects any extensions.

I'm open to the third approach (removing the versioned release schema), but it can't be pursued without a minor version (1.2), as it's a significant change.

cc @mpostelnicu, @duncandewhurst, @kindly, @LindseyAM

### PR DESCRIPTION
**Pull request by jpmckinney**
Notable changes:

* Merging documentation is rewritten
* `tag` is omitWhenMerged #770 
* `Budget.id`, `Identifier.id` and `Classification.id` are versioned #798 
* `buyer.id`, `Tender.procuringEntity.id`, `Transaction.payer.id`, and `Transaction.payee.id` are versioned, using a special `OrganizationReferenceVersionedId` definition #798 
* Add versionId to Identifier and Classification
* All unused `*Unversioned` definitions are removed
* Add a dereferenced release schema

closes #770, closes #798, closes #771

### CODE REVIEW

**Comment by jpmckinney** [comment]
Ready for review - please see linked issues for context.

**Comment by jpmckinney** [comment]
@mpostelnicu The versioned release schema in this pull request should now be corrected.

File: standard/docs/en/schema/changelog.md
Comment (duncandewhurst) [inline standard/docs/en/schema/changelog.md]: PR #806 incorrectly labelled as #808. Also, should the italian translation have it's own bullet point? (currently it follows on from translation and localization policy bullet)

**Comment by jpmckinney** [comment]
@duncandewhurst This PR is still WIP as I need to further clarify merging.md

**Comment by duncandewhurst** [comment]
Ah, OK - let me know when I should restart my review.

**Comment by jpmckinney** [comment]
@duncandewhurst The PR is again ready for review. merging.md was rewritten. I recommend instead reading it at http://standard.open-contracting.org/merge-rules/en/schema/merging/ (once it's built)

File: standard/docs/en/schema/merging.md
Comment (duncandewhurst) [inline standard/docs/en/schema/merging.md]: > If the **input** value is not `null`...

Since a literal value will never be encountered other than as a property of an object, I think this case is already covered by the "If the field in **input** has a value of `null`..." rule in the object value section.

We should either remove the "If the **input** value is not `null`..." clause here, or if repeating it for clarity, we should also repeat what to do if the field is `null`.

File: standard/docs/en/schema/merging.md
Comment (duncandewhurst) [inline standard/docs/en/schema/merging.md]: We could provide a compiled schema in OCDS for the convenience of reimplementations of merging written in something other than python.

There's some work happening in BODS to add generate a dereferenced schema as part of the standard build process which we might be able to re-use.

File: standard/docs/en/schema/merging.md
Comment (duncandewhurst) [inline standard/docs/en/schema/merging.md:15]: It would be good to expand the example (or add another example) demonstrating null-ing for values, objects and arrays.

I don't think we need to do that now, but I added it to the list of suggested examples for OCDS 1.1.5

**Comment by duncandewhurst** [comment]
I reviewed the built version of `merging.md` and suggested that @mrshll1001 reads through the [built page](http://standard.open-contracting.org/merge-rules/en/schema/merging/) too as a fresh pair of eyes, since I already understand merging.

I'll continue reviewing the other changes

File: standard/docs/en/schema/merging.md
Comment (mrshll1001) [inline standard/docs/en/schema/merging.md]: This is a bit jarring in context of the previous paragraphs. We go from "Records are this." to "Records make use of merging" to "Merging is used for compiled and versioned releases and here are details" to "Here's a repeat of the versioned release details".

Recommend merging this content with line 6 to avoid duplication and creating a better flow for the introduction.

File: standard/docs/en/schema/merging.md
Comment (mrshll1001) [inline standard/docs/en/schema/merging.md:128]: This information is critical to provide to users and is well explained here :-) I suggest it be signposted earlier in the document in the introduction. I'll make a comment where I think it would be good to include this.

File: standard/docs/en/schema/merging.md
Comment (mrshll1001) [inline standard/docs/en/schema/merging.md:9]: As per my later comment about essential information, I suggest signposting the technical processes described later in detail.

The bullet points on lines 5 and 6 do a good job of explaining the differences between compiled and versioned releases. We should also indicate at a high level what the implementation of merging will look like.

I suggest something to the effect of "A compiled release is produced by taking the latest version of each field from the releases in a contracting process" and "A versioned release is produced by replacing a field's value with an array of versioned values, which should be in chronological order by `releaseDate`"

File: standard/docs/en/schema/merging.md
Comment (mrshll1001) [inline standard/docs/en/schema/merging.md:87]: Suggest adding a brief explanation of why these fields are discarded. It *should* be obvious to people who know the standard enough to follow the merging advice, but worth addressing explicitly.

File: standard/docs/en/schema/merging.md
Comment (mrshll1001) [inline standard/docs/en/schema/merging.md:102]: Suggest every time "merge routine" is mentioned, it is hyperlinked back to the relevant heading. This supports users who make heavy use of swapping in their short-term memories ;-)

File: standard/docs/en/schema/merging.md
Comment (mrshll1001) [inline standard/docs/en/schema/merging.md:182]: Lot of use of the word *value* here. Might be unavoidable due to the `value` field. I suggest looking at potential rephrasing as it's a bit jarring to read.

File: standard/docs/en/schema/merging.md
Comment (mrshll1001) [inline standard/docs/en/schema/merging.md:84]: Suggest reviewing structure of subsections here to make it easier to find information pertaining to specific types of release (compiled vs versioned).

Current subsections describing specific areas are good, and would be enhanced by a consistent structure ie

```markdown
### Objects
#### In a Compiled Release
....
#### In a Versioned Release
```

Comment (mrshll1001) [reply standard/docs/en/schema/merging.md:84]: I did notice that this is somewhat already present in some subsections. I think we could be more consistent.

**Review by duncandewhurst — APPROVED** [review]
Of my comments, only https://github.com/open-contracting/standard/pull/810#discussion_r269912163 is important to resolve before merging.

File: standard/docs/en/schema/changelog.md
Comment (jpmckinney) [reply standard/docs/en/schema/changelog.md]: Fixed in f3c05cdab953293d305fbb6ca86677d0a3db246c

File: standard/docs/en/schema/merging.md
Comment (jpmckinney) [reply standard/docs/en/schema/merging.md]: Removed in fb0aff6

File: standard/docs/en/schema/merging.md
Comment (jpmckinney) [reply standard/docs/en/schema/merging.md]: I added a dereferenced release schema and linked to it in fb0aff6

File: standard/docs/en/schema/merging.md
Comment (jpmckinney) [reply standard/docs/en/schema/merging.md]: Removed in fb0aff6

File: standard/docs/en/schema/merging.md
Comment (jpmckinney) [reply standard/docs/en/schema/merging.md:182]: I don't know how to rephrase, but we can revisit at another time. Replacing "versioned value" with another term would help, but I can't think of one.

File: standard/docs/en/schema/merging.md
Comment (jpmckinney) [reply standard/docs/en/schema/merging.md]: Yeah, that last paragraph was on the original page and I didn't know where to put it. The worked example demonstrates the utility of the versioned release, so I deleted this paragraph in 50ee7f9

File: standard/docs/en/schema/merging.md
Comment (jpmckinney) [reply standard/docs/en/schema/merging.md:9]: Added bullet points in 50ee7f9

File: standard/docs/en/schema/merging.md
Comment (jpmckinney) [reply standard/docs/en/schema/merging.md:87]: Added in 50ee7f9

File: standard/docs/en/schema/merging.md
Comment (jpmckinney) [reply standard/docs/en/schema/merging.md:102]: Done throughout the merge routine section in 50ee7f9

File: standard/docs/en/schema/merging.md
Comment (jpmckinney) [reply standard/docs/en/schema/merging.md:84]: "Discarded fields" is the only section in which both types of releases are mentioned but not segmented visually using indented lists. Fixed in 50ee7f9

**Comment by jpmckinney** [comment]
I've resolved all comments, except two (example to be part of 1.1.5, and repetition of value to be worked on later https://github.com/open-contracting/standard/pull/810#discussion_r270074822) that don't block merging.

**Comment by jpmckinney** [comment]
I added another commit relating to #831, in which all norms are changed to MUST except for the chronological order of versioned values, which was never communicated anywhere before, so it can be at most a SHOULD.

**Review by duncandewhurst — APPROVED** [review]
Changes look good.

I realised that implicit in the ordering part of the merge routine is a requirement that `date` is unique within the scope of a contracting process - I've opened https://github.com/open-contracting/standard/issues/848 to suggest we make that explicit in the schema.

This actually came up as an issue in the Australia data where the publisher was updating the release date for all releases relating to a contracting process whenever anything changed.

**Comment by jpmckinney** [comment]
Thanks! We'll have an opportunity to further refine this page in 1.1.5, so merging.

**Pull request merged**
